lib/inherited_resources/url_helpers.rb

```diff
@@ -184,56 +174,71 @@ def generate_url_and_path_helpers(prefix, name, resource_segments, resource_ivar
       singleton   = self.resources_configuration[:self][:singleton]
       polymorphic = self.parents_symbols.include?(:polymorphic)
 
-      # If it's not a singleton, ivars are not empty, not a collection or
-      # not a "new" named route, we can pass a resource as argument.
-      #
-      unless (singleton && name != :parent) || ivars.empty? || name == :collection || prefix == :new
-        ivars.push "(given_args.first || #{ivars.pop})"
-      end
-
       # In collection in polymorphic cases, allow an argument to be given as a
       # replacemente for the parent.
       #
-      if name == :collection && polymorphic
-        index = ivars.index(:parent)
-        ivars.insert index, "(given_args.first || parent)"
-        ivars.delete(:parent)
-      end
+      parent_index = ivars.index(:parent) if polymorphic
 
-      # When polymorphic is true, the segments must be replace by :polymorphic
-      # and ivars should be gathered into an array, which is compacted when
-      # optional.
-      #
-      if polymorphic
-        segments = :polymorphic
-        ivars    = "[#{ivars.join(', ')}]"
-        ivars   << '.compact' if self.resources_configuration[:polymorphic][:optional]
+      segments = if polymorphic
+        :polymorphic
+      elsif resource_segments.empty?
+        'root'
       else
-        segments = resource_segments.empty? ? 'root' : resource_segments.join('_')
-        ivars    = ivars.join(', ')
+        resource_segments.join('_')
       end
 
-      prefix = prefix ? "#{prefix}_" : ''
-      ivars << (ivars.empty? ? 'given_options' : ', given_options')
+      ivars = ivars.present? ? Array(ivars) : []
```

Comment:
After all the refactoring `ivars` is never converted to string, so I think this whole line can just go?